internal/controller/instaslice_controller.go

```diff
@@ -105,61 +119,86 @@ func (r *InstasliceReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		}
 	}
 
-	var instasliceList inferencev1alpha1.InstasliceList
-
-	if err := r.List(ctx, &instasliceList, &client.ListOptions{}); err != nil {
-		log.FromContext(ctx).Error(err, "Error listing Instaslice")
-	}
-
 	// failed pods are not deleted by InstaSlice, finalizer is removed so that user can
-	// delete the pod and later InstaSlice will clean up the MIG slice.
+	// delete the pod.
 	if pod.Status.Phase == v1.PodFailed && controllerutil.ContainsFinalizer(pod, finalizerName) {
+		allocationNotFound := true
 		for _, instaslice := range instasliceList.Items {
 			for _, allocation := range instaslice.Spec.Allocations {
-				found, result, err := r.processInstasliceAllocation(ctx, instaslice.Name, string(pod.UID), allocation, req)
-				if err != nil {
-					return result, err
-				}
-				if found {
-					return ctrl.Result{}, nil
+				if pod.UID == types.UID(allocation.PodUUID) {
+					allocationNotFound = false
+					if allocation.Allocationstatus == inferencev1alpha1.AllocationStatusCreating {
+						return ctrl.Result{RequeueAfter: 2 * time.Second}, nil
+					}
+					if allocation.Allocationstatus == inferencev1alpha1.AllocationStatusCreated || allocation.Allocationstatus == inferencev1alpha1.AllocationStatusUngated {
+						resultDeleting, errInDeleting := r.setInstasliceAllocationToDeleting(ctx, instaslice.Name, string(pod.UID), allocation)
+						if errInDeleting != nil {
+							return resultDeleting, nil
+						}
+						// return and rely on daemonset to se allocation status to created
+						// this will cause podmap function to wakeup pod and perform clean up
+						return ctrl.Result{}, nil
+					}
+					if allocation.Allocationstatus == inferencev1alpha1.AllocationStatusDeleted {
+						resultRemove, errInRemove := r.removeInstasliceAllocation(ctx, instaslice.Name, allocation)
+						if errInRemove != nil {
+							return resultRemove, nil
```

Comment:
errInRemove is not propagated here.

Reply:
We requeue request, this is a transient error, I do not return error to avoid the requeue immediately, InstaSlice CR is updated by two parties in the system

Reply:
Ok. Let's log the err instead then.

Reply:
```golang
return ctrl.Result{RequeueAfter: requeueDelay, Requeue: true}, err
```
would that work?

Reply:
Thanks. I prefer logging the error. There is a discussion about logging levels, and noisy or false positive errors could be logged at the debug level.

Reply:
Returning nil, err will use the default exponential backoff... Seems reasonable to use the default backoff.

https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.19.0/pkg/reconcile#TypedReconciler

Reply:
The goal is to perform the retry as soon as possible, not removing allocations soon will cause latency issues in scheduling the pod and will bloat the InstaSlice object which again causes consistency issues when both the controller update InstaSlice object.